tmt/steps/provision/__init__.py

```diff
@@ -59,6 +59,235 @@ class CheckRsyncOutcome(enum.Enum):
     INSTALLED = 'installed'
 
 
+class GuestPackageManager(enum.Enum):
+    DNF = 'dnf'
+    YUM = 'yum'
+    RPM_OSTREE = 'rpm-ostree'
+
+
+T = TypeVar('T')
+
+
+@dataclasses.dataclass
+class GuestFacts(tmt.utils.SerializableContainer):
+    """
+    Contains interesting facts about the guest.
+
+    Inspired by Ansible or Puppet facts, interesting guest facts tmt
+    discovers while managing the guest are stored in this container,
+    plus the code performing the discovery of these facts.
+    """
+
+    #: Set to ``True`` by the first call to :py:meth:`sync`.
+    in_sync: bool = False
+
+    arch: Optional[str] = None
+    distro: Optional[str] = None
+    kernel_release: Optional[str] = None
+    package_manager: Optional[GuestPackageManager] = field(
+        # cast: since the default is None, mypy cannot infere the full type,
+        # and reports `package_manager` parameter to be `object`.
+        default=cast(Optional[GuestPackageManager], None),
+        serialize=lambda package_manager: package_manager.value if package_manager else None,
+        unserialize=lambda raw_value: GuestPackageManager(raw_value) if raw_value else None)
+
+    has_selinux: Optional[bool] = None
+
+    os_release_content: Dict[str, str] = field(default_factory=dict)
+    lsb_release_content: Dict[str, str] = field(default_factory=dict)
+
+    # TODO nothing but a fancy helper, to check for some special errors that
+    # may appear this soon in provisioning. But, would it make sense to put
+    # this detection into the `GuestSsh.execute()` method?
+    def _execute(
+            self,
+            guest: 'Guest',
+            command: Command) -> Optional[tmt.utils.CommandOutput]:
+        """
+        Run a command on the given guest.
+
+        On top of the basic :py:meth:`Guest.execute`, this helper is able to
+        detect a common issue with guest access. Facts are the first info tmt
+        fetches from the guest, and would raise the error as soon as possible.
+
+        :returns: command output if the command quit with a zero exit code,
+            ``None`` otherwise.
+        :raises tmt.units.GeneralError: when logging into the guest fails
+            because of a username mismatch.
+        """
+
+        try:
+            return guest.execute(command, silent=True)
+
+        except tmt.utils.RunError as exc:
+            if exc.stdout and 'Please login as the user' in exc.stdout:
+                raise tmt.utils.GeneralError(f'Login to the guest failed.\n{exc.stdout}') from exc
+
+        return None
+
+    def _fetch_keyval_file(self, guest: 'Guest', filepath: Path) -> Dict[str, str]:
+        """
```

Comment:
How about shlex.split or similar? If this is intended for os-release files then it can contain also comments.
See https://www.freedesktop.org/software/systemd/man/os-release.html

Reply:
Fixed in 76a8f4d3. The linked docs even provide a Python example.